### ISSUE TITLE
RemoteId without discovery?

### ISSUE THREAD

**Issue by i8beef** [jlynch630/Harmony.NET#4]
If I know the IP address, is there a way to instantiate the Hub without knowing what the RemoteId is? Or is there another way for us to get that?

**Comment by i8beef** [comment]
Hmm thats actually exactly what I tried there, but I am not discovering ANY of my hubs on my local network with that. Any ideas on why discovery would fail?

**Comment by i8beef** [comment]
Specifically, I'm trying to do this:

```
        /// <summary>
        /// Uses discovery to find the hub on the network.
        /// </summary>
        /// <param name="harmonyIp">Expected IP address of hub.</param>
        /// <returns>A <see cref="HarmonyHub.HubInfo" />.</returns>
        private static async Task<HarmonyHub.HubInfo> DiscoverHubInfoAsync(string harmonyIp)
        {
            var discoveryService = new HarmonyHub.DiscoveryService();
            var waitCancellationToken = new CancellationTokenSource();
            HarmonyHub.HubInfo hubInfo = null;
            discoveryService.HubFound += (sender, e) =>
            {
                if (e.HubInfo.IP == harmonyIp)
                {
                    Log.Logger.Information("Found hub " + e.HubInfo.IP + " (" + e.HubInfo.RemoteId + ")");
                    hubInfo = e.HubInfo;
                    waitCancellationToken.Cancel();
                }
            };

            // Run discovery for 30 seconds
            discoveryService.StartDiscovery();

            try
            {
                await Task.Delay(30000, waitCancellationToken.Token);
            }
            catch (OperationCanceledException)
            {
                // token has been canceled. exit
            }

            discoveryService.StopDiscovery();

            return hubInfo;
        }
```

**Comment by jlynch630** [comment]
What is the output of `ipconfig` or `ifconfig` on your local machine? Your sample works on my end so it's unlikely a problem with your code.

**Comment by i8beef** [comment]
Windows IP Configuration

   Host Name . . . . . . . . . . . . : REDACTED
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : REDACTED

Ethernet adapter Local Area Connection:

   Connection-specific DNS Suffix  . : REDACTED
   Description . . . . . . . . . . . : Intel(R) 82579V Gigabit Network Connection
   Physical Address. . . . . . . . . : 14-DA-E9-BC-EC-2E
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::ace1:47c0:bb21:69a3%6(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.1.135(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : Wednesday, December 12, 2018 9:15:37 PM
   Lease Expires . . . . . . . . . . : Thursday, December 20, 2018 2:09:27 PM
   Default Gateway . . . . . . . . . : 192.168.1.1
   DHCP Server . . . . . . . . . . . : 192.168.1.1
   DHCPv6 IAID . . . . . . . . . . . : 236247785
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-16-30-93-98-14-DA-E9-BC-EC-2E
   DNS Servers . . . . . . . . . . . : 192.168.1.1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Bluetooth Network Connection:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Bluetooth PAN HelpText
   Physical Address. . . . . . . . . : 00-26-83-33-56-8A
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter vEthernet (Default Switch):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter
   Physical Address. . . . . . . . . : D6-15-30-CB-64-BB
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::80d4:44f2:d724:8357%9(Preferred)
   IPv4 Address. . . . . . . . . . . : 172.27.92.161(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.240
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 251663709
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-16-30-93-98-14-DA-E9-BC-EC-2E
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Disabled

**Comment by i8beef** [comment]
Got it. Firewall is blocking. Disabling it temporarily let it start up.

**Comment by jlynch630** [comment]
Ah, well there you go. I'll update the repository soon to try the local broadcast address (`192.168.1.255`) first. That should let you keep your firewall

**Comment by ehendrix23** [comment]
FYI, figured out how to get remoteId for 1 specific HUB.

Open a HTTP request (POST) to port 8088 with what seems to be following JSON:
{"id ": 123,"cmd":"connect.discoverinfo?get","params":}}}

What is then returned is a JSON with what seems to be the same information as from discovery.

**Comment by i8beef** [comment]
If that works, can we look at adding that approach as a Hub constructor like `new Hub(ipAddress)` and it just goes and gets the remote id?

**Comment by jlynch630** [comment]
I can't actually reproduce that @ehendrix23 on my local Hub. Regardless, that sounds like something that was definitely *not* intended by Logitech and will refrain from implementing it, at least until I can find if Logitech ever uses it anywhere

**Comment by i8beef** [comment]
Can't reproduce with `curl` locally either... You missing something in your capture there?

**Comment by ehendrix23** [comment]
I was able to determine this by capturing the data going from the App to the Hub.
On the App, go to Switch Hub. Click on ? and then select "Enter IP address". Enter IP address then and click Connect. It is from there that I captured it.

I also was just able to test this and confirm it works.
URL: http://<ip address>:8080/
Headers:
  Origin = http://localhost.nebula.myharmony.com
  Content-Type = application/json
  Accept = application/json
  accept-Charset = utf-8

JSON request:
  {
        "id ": 124,
        "cmd": "connect.discoveryinfo?get",
        "params": {}
    }

Note, I missed the y in discoveryinfo in my earlier post. :-)

Response back is a json object.

**Comment by i8beef** [comment]
Confirmed. Its the Origin header that's necessary to make it accept. It won't answer without that.

**Comment by jlynch630** [comment]
I was able to get it to work, but the request took over a minute to fulfill. Is that just me?

**Comment by i8beef** [comment]
Mine comes back immediately. I did edit the payload to all be one line to make `curl` happy though. Also there's a space in the supplied "id " there that I removed.

**Comment by ehendrix23** [comment]
FYI, also just checked and it seems to Hub will notify when an activity is started. I was monitoring traffic and just pretty much saw ping/pong web socket traffic. I then started an activity and Hub sent a notify message that included activityId. At that time seems to be a bit back and forth going on.
Getting late, will be working on that part (together with everything else) tomorrow and following days/weeks. :-)

**Comment by ehendrix23** [comment]
@jlynch630 Mine comes back right away.

@i8beef It all goes as 1 request when sending it. What I captured had that space in "id " there hence left it. :-)

**Comment by jlynch630** [comment]
Yeah I'll commit what I have for now, but am going to hit the hay as well @ehendrix23

**Comment by sirmeili** [comment]
> FYI, also just checked and it seems to Hub will notify when an activity is started. I was monitoring traffic and just pretty much saw ping/pong web socket traffic. I then started an activity and Hub sent a notify message that included activityId. At that time seems to be a bit back and forth going on.
> Getting late, will be working on that part (together with everything else) tomorrow and following days/weeks. :-)

A bit off topic to this issue, but since you brought it up:

When starting an activity you should get 4 messages back from the hub. If you have an activity that takes a long time to start you'll see them spaced out more. You'll get a message for when the hub is starting to start the activity, one when it's finished starting the activity, and one when the hub has updated it's internal activityID. the 4th one seems to be another update which now includes the activity ID in the list of active activities (weird, because you can only have 1 active activity as far as I know).

These are used I believe to control the states of the phone app (show the connecting throbber, etc.). I actually use these different states in my implementation.

**Comment by jlynch630** [comment]
The WebSocket should actually also return messages which are unused by the app which indicates the progress of starting the activity, i.e. number of devices turned on and total number of devices. Do you not get these? They're under the "startActivity" or "helpDiscretes" commands

**Comment by sirmeili** [comment]
I do not think I do, but that would be really useful. I'll do more research tonight. Another boring day of meetings ;(

**Comment by ehendrix23** [comment]
FYI, I just pushed updates for Python library used by HASS. So far so good. :-)

https://github.com/iandday/pyharmony/pull/21
